spring-modulith-events/spring-modulith-events-mongodb/src/main/java/org/springframework/modulith/events/mongodb/MongoDbEventPublicationRepository.java

```diff
@@ -0,0 +1,155 @@
+ */
+package org.springframework.modulith.events.mongodb;
+
+import org.springframework.data.domain.Sort;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import org.springframework.data.mongodb.core.query.Criteria;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.data.util.TypeInformation;
+import org.springframework.modulith.events.CompletableEventPublication;
+import org.springframework.modulith.events.EventPublication;
+import org.springframework.modulith.events.EventPublicationRepository;
+import org.springframework.modulith.events.PublicationTargetIdentifier;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import lombok.EqualsAndHashCode;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Repository to store {@link EventPublication}s in a MongoDB.
+ *
+ * @author Björn Kieling
+ * @author Dmitry Belyaev
+ */
+public class MongoDbEventPublicationRepository implements EventPublicationRepository {
+
+	private final MongoTemplate mongoTemplate;
+
+	public MongoDbEventPublicationRepository(MongoTemplate mongoTemplate) {
+		this.mongoTemplate = mongoTemplate;
+	}
+
+	@Override
+	public EventPublication create(EventPublication publication) {
+		mongoTemplate.save(domainToDocument(publication));
+		return publication;
+	}
+
+	@Override
+	public EventPublication update(CompletableEventPublication publication) {
+		findDocumentsByEventAndTargetIdentifier(publication.getEvent(), publication.getTargetIdentifier())
+				.stream()
+				.findFirst()
+				.ifPresent(document -> publication.getCompletionDate()
+						.ifPresent(completionDate -> {
+							document.setCompletionDate(completionDate);
+							mongoTemplate.save(document);
+						}));
+		return publication;
+	}
+
+	@Override
+	public List<EventPublication> findIncompletePublications() {
+		var query = Query.query(Criteria.where("completionDate").isNull());
+		return mongoTemplate.find(query, MongoDbEventPublication.class).stream() //
+				.map(this::documentToDomain) //
+				.collect(Collectors.toList());
+	}
+
+	@Override
+	public Optional<EventPublication> findByEventAndTargetIdentifier(
+			Object event, PublicationTargetIdentifier targetIdentifier) {
+
+		var documents = findDocumentsByEventAndTargetIdentifier(event, targetIdentifier);
+		var results = documents
+				.stream() //
+				.map(this::documentToDomain) //
+				.toList();
+
+		// if there are several events with exactly the same payload we return the oldest one first
+		return results.isEmpty() ? Optional.empty() : Optional.of(results.get(0));
+	}
+
+	private List<MongoDbEventPublication> findDocumentsByEventAndTargetIdentifier(
+			Object event, PublicationTargetIdentifier targetIdentifier) {
+
+		// we need to enforce writing of the type information
+		Object eventAsMongoType = mongoTemplate.getConverter().convertToMongoType(event, TypeInformation.of(Object.class));
+		var criteria = List.of( //
```

Comment:
Building up a `Criteria` via simple `….and(…)` concatenation is sufficient here. We'd only need the `….andOperator(…)` if we were to define different constraints on the same key, which we aren't.

Reply:
Done